read_buffer/src/lib.rs

```diff
@@ -257,18 +260,55 @@ impl Database {
     ///
     /// This method might be deprecated in the future, replaced by a call to
     /// `read_aggregate_window` with a `window` of `0`.
-    pub fn read_aggregate(
+    pub fn read_aggregate<'input>(
         &self,
         partition_key: &str,
-        table_name: &str,
+        table_name: &'input str,
         chunk_ids: &[u32],
         predicate: Predicate,
-        group_columns: ColumnSelection<'_>,
-        aggregates: Vec<(ColumnName<'_>, AggregateType)>,
-    ) -> Result<ReadAggregateResults> {
-        Err(Error::UnsupportedOperation {
-            msg: "`read_aggregate` not yet implemented".to_owned(),
-        })
+        group_columns: ColumnSelection<'input>,
+        aggregates: Vec<(ColumnName<'input>, AggregateType)>,
+    ) -> Result<ReadAggregateResults<'input, '_>> {
+        match self.partitions.get(partition_key) {
+            Some(partition) => {
+                let mut chunks = vec![];
+                for chunk_id in chunk_ids {
+                    let chunk = partition
+                        .chunks
+                        .get(chunk_id)
+                        .context(ChunkNotFound { id: *chunk_id })?;
+
+                    ensure!(chunk.has_table(table_name), TableNotFound { table_name });
+
+                    chunks.push(
+                        partition
+                            .chunks
+                            .get(chunk_id)
+                            .ok_or_else(|| Error::ChunkNotFound { id: *chunk_id })?,
+                    )
+                }
+
+                for (_, agg) in &aggregates {
+                    match agg {
+                        AggregateType::First | AggregateType::Last => {
+                            return Err(Error::UnsupportedAggregate { agg: *agg });
+                        }
+                        _ => {}
+                    }
+                }
+
+                Ok(ReadAggregateResults::new(
+                    chunks,
+                    table_name,
+                    predicate,
+                    group_columns,
+                    aggregates,
+                ))
+            }
+            None => Err(Error::PartitionNotFound {
+                key: partition_key.to_owned(),
+            }),
```

Comment:
You can also probably write this as follows, if you are inclined to want to use more Sanfu magic:

```suggestion
            None => PartitionNotFound {
                key: partition_key,
            }.fail(),
```